openedx/core/djangoapps/content_staging/views.py

```diff
@@ -0,0 +1,151 @@
+"""
+REST API views for content staging
+"""
+import logging
+
+from django.db.transaction import atomic
+from django.http import HttpResponse
+from django.shortcuts import get_object_or_404
+import edx_api_doc_tools as apidocs
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import UsageKey
+from opaque_keys.edx.locator import CourseLocator
+from rest_framework.exceptions import NotFound, PermissionDenied, ValidationError
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from common.djangoapps.student.auth import has_studio_read_access
+
+from openedx.core.lib.api.view_utils import view_auth_classes
+from xmodule import block_metadata_utils
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.exceptions import ItemNotFoundError
+
+from .block_serializer import XBlockSerializer
+from .models import StagedContent, UserClipboard
+from .serializers import UserClipboardSerializer, PostToClipboardSerializer
+from .tasks import delete_expired_clipboards
+
+log = logging.getLogger(__name__)
+
+
+@view_auth_classes(is_authenticated=True)
+class StagedContentOLXEndpoint(APIView):
+    """
+    API Endpoint to get the OLX of any given StagedContent.
+    """
+
+    def get(self, request, id):  # pylint: disable=redefined-builtin
+        """
+        Get the OLX of the given StagedContent object.
+        """
+        staged_content = get_object_or_404(StagedContent, pk=id)
+        if staged_content.user.id != request.user.id:
+            raise PermissionDenied("Users can only access their own staged content")
+        if staged_content.status != StagedContent.Status.READY:
+            # If the status is LOADING, the OLX may not be generated/valid yet.
+            # If the status is ERROR or EXPIRED, this row is no longer usable.
+            raise NotFound("The requested content is not available.")
+        return HttpResponse(staged_content.olx, headers={
+            "Content-Type": f"application/vnd.openedx.xblock.v1.{staged_content.block_type}+xml",
+            "Content-Disposition": f'attachment; filename="{staged_content.olx_filename}"',
+        })
+
+
+@view_auth_classes(is_authenticated=True)
```

Comment:
You might want to decorate this to prevent Django from ever enclosing this in a view-wide implicit transaction (which could later cause concurrency issues with cleanup):

```python
@method_decorator(transaction.non_atomic_requests, name='dispatch')
```

Reply:
Thanks, done.